### PR TITLE
fix validate to admit manually modified sample sheets

### LINKED ISSUE #3288: Add flexibility to the sample sheet validation
## Description

This week a flow cell needed to have manual modifications to its sample sheet, in a way that it failed the implemented validation. This modified sample sheet will be needed when the flow cell is set for re-demultiplexing. However, at the start of the demultiplexing, the automation overwrites any sample sheet (in Housekeeper or in the Flow cell directory) that fails validation with a new one. This means that the modified sample sheet can't be saved in Housekeeper or in the flow cell directory as a normal sheet.

## Suggested solution

We most likely need to allow sample sheets that fail validations for these cases somehow, but it is unclear how.

## This can be closed when

This flow cell can be re-demultiplexed without errors

## Blocked by

*If there are any blocking issues/prs/things in this or other repos. Please link to them.*

### PR DESCRIPTION
**Pull request by diitaz93**
## Description
Fixes #3288 
This PR introduces a distinction between a sample sheet failing validation by structural failures (most likely Bcl2Fastq or other old sample sheets) and failing by content (manually modified sample sheets). For this, two new exceptions replace the `SampleSheetError`: `SampleSheetFormatError` and `SampleSheetContentError`, respectively. Sample sheets failing validation for format will be regenerated as usual, but sample sheets failing for content will not.

### Added

- New exceptions `SampleSheetFormatError` and `SampleSheetContentError`

### Changed

- Replaced all occurrences of `SampleSheetError` for either `SampleSheetFormatError` and `SampleSheetContentError`
- Fixed the try-except clauses to perform different conditions for the two different exceptions
- Sample sheet Error only used in Nextflow code renamed to `NfSampleSheetError`

### How to prepare for test

- [x] Ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] Paxa the environment: `paxa`
- [x] Install on stage (example for Hasta):
    ```shell
    bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_cg -t cg -b fix-sample-sheet-validation -a
    ```

### How to test

- [x] Create a sample sheet when a manually modified sample sheet exists in the sequencing dir, see that the validation fails but the sample sheet is not regenerated
```shell
$ cg -l DEBUG demultiplex samplesheet create 20231108_LH00188_0028_B22F52TLT3 --dry-run
Running cg demultiplex.
Getting a valid sample sheet for flow cell 20231108_LH00188_0028_B22F52TLT3
Instantiating sample sheet API
Instantiating housekeeper api
Initializing Store
Instantiating lims api
Called undefined __fields__ on HousekeeperAPI, please wrap
Set dry run to True
Set force to False
Instantiating IlluminaRunDirectoryData with path /home/proj/stage/sequencing_data/illumina/sequencing-runs/20231108_LH00188_0028_B22F52TLT3
Set sequencing run id to B22F52TLT3
Fetching and validating sample sheet from Housekeeper
Fetch latest version from bundle 22F52TLT3
Fetching files with tags in [22F52TLT3,samplesheet]
Fetching files from version 163391
Sample sheet file for flowcell 22F52TLT3 not found in Housekeeper!
Sample sheet from Housekeeper is not correctly formatted or does not exist, trying sample sheet in sequencing directory
Validating sample sheet
Validating that the sample sheet has all the necessary sections
Looking for index settings in the sample sheet
Found index settings: NovaSeqX
Looking for read and index run cycles in the sample sheet
Validating samples
Order samples by lane
Validate that samples are unique in lane: 1
Sample ACC13169A1 exists multiple times in sample sheet
Validation failed for /home/proj/stage/sequencing_data/illumina/sequencing-runs/20231108_LH00188_0028_B22F52TLT3/SampleSheet.csv. Possibly manually modified sample sheet. Sample sheet will not be re-generated.
```
- [x] Create a sample sheet when a manually modified sample sheet exists in Housekeeper and a correct sample sheet exists in the sequencing run: see that the validation fails but the sample sheet is not regenerated
```shell
Fetching and validating sample sheet for 20231108_LH00188_0028_B22F52TLT3 from Housekeeper
Sample ACC13169A1 exists multiple times in sample sheet
Validation failed for /home/proj/stage/housekeeper-bundles/22F52TLT3/2023-11-08/SampleSheet.csv. Possibly a manually modified sample sheet. Sample sheet will not be re-generated.
```
- [x] Create a sample sheet when a sample sheet with incorrect sections exists in the sequencing dir, see that the validation fails and the sample sheet is regenerated
```shell
Running cg demultiplex.
Getting a valid sample sheet for flow cell 20231108_LH00188_0028_B22F52TLT3
Called undefined __fields__ on HousekeeperAPI, please wrap
Fetching and validating sample sheet for 20231108_LH00188_0028_B22F52TLT3 from Housekeeper
Sample sheet file for flowcell 22F52TLT3 not found in Housekeeper!
Sample sheet from Housekeeper is not correctly formatted or does not exist, trying sample sheet in sequencing directory
No index settings found in sample sheet
Sample sheet from sequencing directory is not correctly formatted or does not exist, creating new sample sheet
Fetching samples from lims for flowcell 22F52TLT3
Constructing sample sheet for the novaseqx flow cell 22F52TLT3
Updating barcode mismatch values for samples in lane 1
Updating barcode mismatch values for samples in lane 2
Updating barcode mismatch values for samples in lane 3
Updating barcode mismatch values for samples in lane 4
Updating barcode mismatch values for samples in lane 5
Updating barcode mismatch values for samples in lane 6
Updating barcode mismatch values for samples in lane 7
Updating barcode mismatch values for samples in lane 8
Creating sample sheet content
Samplesheet passed validation
```
- [x] Demultiplex with a sample sheet modified that fails content validation
```shell
 $ cg -l DEBUG demultiplex sequencing-run 20231108_LH00188_0028_B22F52TLT3 --dry-run
Running cg demultiplex.
Starting demultiplexing of sequencing run 20231108_LH00188_0028_B22F52TLT3
Instantiating sample sheet API
Instantiating housekeeper api
Initializing Store
Instantiating lims api
Called undefined __fields__ on HousekeeperAPI, please wrap
Instantiating demultiplexing api
Called undefined __fields__ on HousekeeperAPI, please wrap
Initialising Process with binary: sbatch
Use base call ['sbatch']
Set environment to stage
DemultiplexingAPI: Set dry run to True
SlurmAPI: Set dry run to True
setting flow cell id to 20231108_LH00188_0028_B22F52TLT3
setting demultiplexed runs dir to /home/proj/stage/sequencing_data/illumina/demultiplexed-runs
Instantiating IlluminaRunDirectoryData with path /home/proj/stage/sequencing_data/illumina/sequencing-runs/20231108_LH00188_0028_B22F52TLT3
Set sequencing run id to B22F52TLT3
Check if demultiplexing is possible for 22F52TLT3
Check if sequencing run is ready for downstream processing
Check if sequencing is done
Sequence is done for sequencing run 22F52TLT3
Check if copy of data from sequence instrument is ready
All data has been transferred for sequencing run 22F52TLT3
Sequencing run 22F52TLT3 is ready for downstream processing
Check if sample sheet exists
Fetch latest version from bundle 22F52TLT3
Fetching files with tags in [22F52TLT3,samplesheet]
Fetching files from version 163391
Validating sample sheet
Validating that the sample sheet has all the necessary sections
Looking for index settings in the sample sheet
Found index settings: NovaSeqX
Looking for read and index run cycles in the sample sheet
Validating samples
Order samples by lane
Validate that samples are unique in lane: 1
Sample ACC13169A1 exists multiple times in sample sheet
Demultiplexing a with a manually modified sample sheet
Would have started demultiplexing 20231108_LH00188_0028_B22F52TLT3
[js.diazboada@hasta:20231108_LH00188_0028_B22F52TLT3] [S_base] $
```

- [x] Tests executed by SD
- [x] "Merge and deploy" approved by VJ
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a

- [x] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan

- [ ] Deployed to stage:
```shell

```
- [ ] Deployed to production:
```shell

```

### CODE REVIEW

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 1]
```diff
@@ -141,34 +148,44 @@ def _use_sample_sheet_from_housekeeper(self, flow_cell: IlluminaRunDirectoryData
         """
         Copy the sample sheet from Housekeeper to the flow cell directory if it exists and is valid.
         """
-        try:
-            sample_sheet_path: Path = self.hk_api.get_sample_sheet_path(flow_cell.id)
-        except HousekeeperFileMissingError:
-            raise SampleSheetError(
-                f"Sample sheet for flow cell {flow_cell.id} does not exist in Housekeeper"
-            )
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/api.py:149]: Are we ok with removing this `try-except`?

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 2]
```diff
@@ -217,21 +234,34 @@ def get_or_create_sample_sheet(self, flow_cell_name: str) -> None:
         housekeeper or creating it if there is not a valid sample sheet.
         """
         flow_cell: IlluminaRunDirectoryData = self._get_flow_cell(flow_cell_name)
-        LOG.info("Fetching and validating sample sheet from Housekeeper")
+        LOG.info(f"Fetching and validating sample sheet for {flow_cell_name} from Housekeeper")
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/api.py:237]: Debug imo
```suggestion
        LOG.debug(f"Fetching and validating sample sheet for {flow_cell_name} from Housekeeper")
```

File: cg/apps/demultiplex/sample_sheet/api.py [hunk 3]
```diff
@@ -217,21 +234,34 @@ def get_or_create_sample_sheet(self, flow_cell_name: str) -> None:
...
+            LOG.warning(
+                "Sample sheet from Housekeeper is not correctly formatted or does not exist, "
+                "trying sample sheet in sequencing directory"
             )
         try:
             self._use_flow_cell_sample_sheet(flow_cell)
             return
-        except SampleSheetError:
+        except SampleSheetContentError:
+            LOG.warning(
```
Comment (Vince-janv) [inline cg/apps/demultiplex/sample_sheet/api.py:256]: Why is `SampleSheetContentError` excepted twice in this function? It makes it very hard to understand what's going on

File: cg/exc.py [hunk 4]
```diff
@@ -192,6 +192,14 @@ class SampleSheetError(CgError):
     """Raised when something is wrong with the sample sheet."""
 
 
+class SampleSheetContentError(CgError):
+    """Raised when something is wrong with the sample sheet content."""
+
+
+class SampleSheetFormatError(CgError):
+    """Raised when something is wrong with the sample sheet format."""
```
Comment (Vince-janv) [inline cg/exc.py:200]: You could create a parent class `SampleSheetError` if you need to catch both of these

**Review by Vince-janv — APPROVED** [review]
Looks good 👌

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 1 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:149]: yes, the purpose of this one was to translate the error from a HousekeeperFileMissingError to a SampleSheetError. All CgErrors are caught later

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 3 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:256]: We first check the content of the sample sheet in housekeeper, check what kind of error is raising and depending on the type of error we would need to do the same for the sample sheet in the sequencing dir

File: cg/exc.py (hunk 4 above)
Comment (diitaz93) [reply cg/exc.py:200]: No, I don't really need to catch both

File: cg/apps/demultiplex/sample_sheet/api.py (hunk 2 above)
Comment (diitaz93) [reply cg/apps/demultiplex/sample_sheet/api.py:237]: fixed

**Pull request merged**